synapse/lib/ast.py

```diff
@@ -2285,42 +2285,78 @@ class Edit(Oper):
     pass
 
 class EditParens(Edit):
+
     async def run(self, runt, genr):
 
         nodeadd = self.kids[0]
         assert isinstance(nodeadd, EditNodeAdd)
 
-        if not nodeadd.isruntsafe(runt):
-            mesg = 'First node add in edit parentheses must not be dependent on incoming nodes'
-            raise s_exc.StormRuntimeError(mesg=mesg)
+        # create an isolated generator for the add vs edit
+        if nodeadd.isruntsafe(runt):
 
-        # Luke, let the nodes flow through you
-        async for item in genr:
-            yield item
+            # Luke, let the (node,path) tuples flow through you
+            async for item in genr:
+                yield item
 
-        # Run the opers once with no incoming nodes
-        genr = agen()
+            # isolated runtime stack...
+            genr = agen()
+            for oper in self.kids:
+                genr = oper.run(runt, genr)
 
-        for oper in self.kids:
-            genr = oper.run(runt, genr)
+            async for item in genr:
+                yield item
 
-        async for item in genr:
-            yield item
+        else:
+
+            # do a little genr-jig.
+            async for node, path in genr:
+
+                yield node, path
+
+                async def editgenr():
+                    async for item in nodeadd.addFromPath(path):
+                        yield item
+
+                fullgenr = editgenr()
+                for oper in self.kids[1:]:
+                    fullgenr = oper.run(runt, fullgenr)
+
+                async for item in fullgenr:
+                    yield item
 
 class EditNodeAdd(Edit):
 
+    def prepare(self):
+
+        oper = self.kids[1].value()
+        self.name = self.kids[0].value()
+
+        self.form = self.core.model.form(self.name)
+        if self.form is None:
+            raise s_exc.NoSuchForm(name=self.name)
+
+        self.excignore = (s_exc.BadTypeValu, s_exc.BadPropValu) if oper == '?=' else ()
+
     def isruntsafe(self, runt):
         return self.kids[2].isRuntSafe(runt)
 
-    async def run(self, runt, genr):
+    async def addFromPath(self, path):
 
-        name = self.kids[0].value()
-        oper = self.kids[1].value()
-        excignore = (s_exc.BadTypeValu, s_exc.BadPropValu) if oper == '?=' else ()
+        valu = await self.kids[2].compute(path)
 
-        form = runt.snap.model.forms.get(name)
+        for valu in self.form.type.getTypeVals(valu):
+            try:
+                newn = await path.runt.snap.addNode(self.name, valu)
+            except self.excignore:
+                pass
+            else:
+                yield newn, path.runt.initPath(newn)
+
+    async def run(self, runt, genr):
+
+        form = runt.snap.model.forms.get(self.name)
         if form is None:
-            raise s_exc.NoSuchForm(name=name)
+            raise s_exc.NoSuchForm(name=self.name)
```

Comment:
This is duplicate of the work done in prepare()

Reply:
Yep, thanks!  Moved it to prepare and forgot to cull